cmd/ztest/ztest.c

```diff
@@ -7052,6 +7119,379 @@ ztest_import(ztest_shared_t *zs)
+	 * temporary datasets which wouldn't otherwise get replayed, which
+	 * can trigger failures when attempting to offline a SLOG in
+	 * ztest_fault_inject().
+	 */
+	(void) dmu_objset_find(ztest_opts.zo_pool, ztest_replay_zil_cb,
+	    NULL, DS_FIND_CHILDREN);
+
+	if (ztest_opts.zo_raidz_expand_test != 0 &&
+	    ztest_opts.zo_raidz_expand_test < UINT64_MAX) {
+		desreflow = ztest_opts.zo_raidz_expand_test;
+		/*
+		 * Set the reflow to pause at the desired offset
+		 */
+		zfs_raidz_expand_max_offset_pause = desreflow;
+		/*
+		 * In here on first pass of test only.
+		 */
+		if (ztest_opts.zo_verbose >= 1) {
+			(void) printf("running raidz expansion test,"
+			    " killing when offset %llu of reflow reached\n",
+			    (u_longlong_t)desreflow);
+			if (ztest_opts.zo_verbose > 1) {
+				/* XXX - pause to allow debugger attach */
+				(void) printf(
+				    "our pid is %d, pausing for 10 seconds\n",
+				    getpid());
+				sleep(10);
+			}
+		}
+		/*
+		 * Put some data in the pool and then attach a vdev to initiate
+		 * reflow.
+		 */
+		/*
+		 * Kick off all the I/O generators that run in parallel.
+		 */
+		for (t = 0; t < ztest_opts.zo_threads; t++) {
+			if (t < ztest_opts.zo_datasets &&
+			    ztest_dataset_open(t) != 0) {
+				umem_free(run_threads, ztest_opts.zo_threads *
+				    sizeof (kthread_t *));
+				return;
+			}
+
+			run_threads[t] = thread_create(NULL, 0,
+			    ztest_rzx_thread,
+			    (void *)(uintptr_t)t, 0, NULL, TS_RUN | TS_JOINABLE,
+			    defclsyspri);
+		}
+
+		/*
+		 * Wait a while for I/O to put some data in the pool
+		 * XXX- add an option to  specify if we wait for I/O to quiesce
+		 */
+		for (i = 0; i < 60; i++) {
+			txg_wait_synced(spa_get_dsl(spa), 0);
+			(void) poll(NULL, 0, 1000);
+		}
+
+		rzvd = spa->spa_root_vdev->vdev_child[0];
+		ASSERT(rzvd->vdev_ops == &vdev_raidz_ops);
+		/*
+		 * get size of a child of the raidz group
+		 */
+		cvd = rzvd->vdev_child[0];
+
+		csize = vdev_get_min_asize(cvd);
+		csize += csize / 10; /* make sure device is a bit bigger */
+		/*
+		 * Path to vdev to be attached
+		 */
+		(void) snprintf(newpath, MAXPATHLEN, ztest_dev_template,
+		    ztest_opts.zo_dir, ztest_opts.zo_pool, rzvd->vdev_children);
+		/*
+		 * Build the nvlist describing newpath.
+		 */
+		root = make_vdev_root(newpath, NULL, NULL, csize, ashift, NULL,
+		    0, 0, 1);
+		/*
+		 * Now attach the vdev to the raidz so it will expand
+		 */
+		if (ztest_opts.zo_verbose >= 1) {
+			(void) printf("expanding raidz\n");
+		}
+		error = spa_vdev_attach(spa, rzvd->vdev_guid, root, B_FALSE,
+		    B_FALSE);
+		nvlist_free(root);
+		if (error != 0) {
+			fatal(0, "raidz expand: attach (%s %llu) returned %d",
+			    newpath, csize, error);
+		}
+
+	} else { /* check the pool is healthy */
+		/*
+		 * Set pool check done flag, main program will run azdb check of
+		 * the pool when we exit.
+		 */
+		ztest_shared_opts->zo_raidz_expand_test = (UINT64_MAX - 1);
```

Comment:
might make sense to have macros for the special values of this shared variable.

Reply:
good idea, will fix